feat: New Dark Scheme Part 2

alifafaruk opened the pull request:
Dark mode support 
fixes #2407

@lucasortizny @Jen-Lopez and I worked on the dark mode feature for the website. We implemented a toggle button, where the user can turn dark mode on/off. We did a [pull request](https://github.com/openfoodfacts/openfoodfacts-server/pull/3316) earlier, but it had multiple merge conflicts that we weren't able to fix. So we closed that. We fixed the issues on our end and now it works!
 
<img width="1440" alt="Screen Shot 2020-04-24 at 5 00 40 PM" src="https://user-images.githubusercontent.com/47068889/80256680-2fc0f500-864d-11ea-8f6f-8e8ff591c780.png">

<img width="1440" alt="Screen Shot 2020-04-24 at 12 58 42 PM" src="https://user-images.githubusercontent.com/47068889/80256388-a6112780-864c-11ea-9153-e450e93c4639.png">

stephanegigandet commented:
Hi @alifafaruk ! Thank you for the PR!

The dark mode looks very good, but instead of adding a site specific toggle, I think it would make much more sense to just use the CSS prefers-color-scheme, as @teolemon suggested in the bug #2407 

That way we can avoid the 2 javascript scripts that will block the page rendering for all users until they load, and we can have a pure CSS solution. 

Also it would be great to put the CSS styles inside the SCSS ( in the /scss/ directory) so that they get processed along with the rest of the CSS styles.

cc @hangy

teolemon commented:
Here's the Open Food Facts logo in dark mode: https://github.com/openfoodfacts/openfoodfacts-server/blob/3d1a30516a663d1dedad0733fac66d9d93a249c8/html/images/misc/openfoodfacts-logo-no-tagline-dark-mode.svg

GendelfLugansk commented on html/js/preDarkMode.js:
```diff
@@ -0,0 +1,10 @@
+// Before page renders, check if page is in darkmode.
+// Otherwise, it will load regular styles before dark style.
+let dark = localStorage.getItem('darkMode');
+var sheet = document.styleSheets[3];
```
The same issue as in `darkMode.js`

GendelfLugansk commented:
It seems like `prefers-color-scheme` requires user to activate dark theme on OS level. Correct me if I'm wrong. In this case I'm not sure if it's better

cc @stephanegigandet @teolemon @hangy

GendelfLugansk commented:
Red text on blue buttons is hard to read, imo. At least on my screen

![screenshot](https://screenshots.gennady.pp.ua/20200503_05fe15403.png)

White text like this is probably a bit too bright. It can bee seen in right sidebar and on some pages with text (like Contribute and Discover)

![screenshot](https://screenshots.gennady.pp.ua/20200503_35267fbc5.png)

That black text (if user is logged in) is barely visible

![screenshots](https://screenshots.gennady.pp.ua/20200503_ede07045d.png)

This block on the right on [category page](http://world.productopener.localhost/category/dairies)

![screenshot](https://screenshots.gennady.pp.ua/20200503_56080a8bf.png)

Some text on [product page](http://world.productopener.localhost/product/3960051050102/beurre-du-haut-jura-les-moussieres) is black and unreadable

![screenshot](https://screenshots.gennady.pp.ua/20200503_9e7c10334.png)

Nutrition table on the same product page is light. Not critical but could be darker, imo

![screenshot](https://screenshots.gennady.pp.ua/20200503_5899fed16.png)

In footer, black text

![screenshot](https://screenshots.gennady.pp.ua/20200503_468eb8df7.png)

Many regions of product edit page are too light-colored, imo

![screenshot](https://screenshots.gennady.pp.ua/20200503_3283454cf.png)
![screenshot](https://screenshots.gennady.pp.ua/20200503_e67033a07.png
![screenshot](https://screenshots.gennady.pp.ua/20200503_8e8d9db30.png)

On Graphs and Maps page, graph has white background

![screenshot](https://screenshots.gennady.pp.ua/20200503_d5c294bb1.png)

GendelfLugansk commented:
Also, it seems like it [might be possible](https://levelup.gitconnected.com/share-localstorage-sessionstorage-between-different-domains-eb07581e9384) to share "dark mode" setting across different domains but this might be complex to implement. I think that would require changes to perl code, docker config and nginx/apache config on production site. I don't have perl expertise to give more detailed advice.

hangy commented:
> It seems like `prefers-color-scheme` requires user to activate dark theme on OS level. Correct me if I'm wrong. In this case I'm not sure if it's better

It sounds like a good idea to default to the dark mode, if someone has selected it, though. [This guide](https://web.dev/prefers-color-scheme/#loading-strategy) explains that the media query can also be done from JS.

> Also, it seems like it [might be possible](https://levelup.gitconnected.com/share-localstorage-sessionstorage-between-different-domains-eb07581e9384) to share "dark mode" setting across different domains but this might be complex to implement. I think that would require changes to perl code, docker config and nginx/apache config on production site.

Changing some JavaScript should suffice, if I understand the blog post correctly. However, I wonder if we need `localStorage` for this. To no waste too much cookie storage, one could just set a `d=1` cookie to store enabled/changed dark mode setting in a cross-subdomain cookie.

hangy commented on html/js/darkMode.js:
```diff
@@ -0,0 +1,30 @@
+let darkMode = localStorage.getItem('darkMode');
+const darkToggle = document.getElementById("switch");
+
+const enableDarkMode = () => {
```
Do you have an idea about browser compatibility? We don't run this JS through a transpiler, so using `const` and lambda expressions could limit dark mode to some modern browsers.

GendelfLugansk replied:
[Const](https://caniuse.com/#feat=const) should be fine. [Arrow functions](https://caniuse.com/#feat=arrow-functions) are not compatible with IE and Opera Mini. I guess that's ok

hangy commented on lib/ProductOpener/Display.pm:
```diff
@@ -6090,7 +6090,13 @@ $options{favicons}
 <link rel="search" href="$formatted_subdomain/cgi/opensearch.pl" type="application/opensearchdescription+xml" title="$Lang{site_name}{$lang}">
 $header
+<!-- Dark Mode -->
+<link rel="stylesheet" href="$static_subdomain/css/darkMode.css">
```
This basically makes dark mode the default - Might introduce unwanted behaviour if JavaScript is disabled. cc @stephanegigandet

stephanegigandet replied:
Dark mode should not be the default.

GendelfLugansk replied:
I didn't test in all browsers but in Chromium adding `disabled="true"` to `link` disables stylesheet. I think this should work in other browsers as well. JS code of this PR is based on `disabled` property and [MDN says](https://developer.mozilla.org/en-US/docs/Web/API/StyleSheet/disabled) it's supported everywhere except IE (unknown)

VaiTon replied:
> 
> 
> I didn't test in all browsers but in Chromium adding `disabled="true"` to `link` disables stylesheet. I think this should work in other browsers as well. JS code of this PR is based on `disabled` property and [MDN says](https://developer.mozilla.org/en-US/docs/Web/API/StyleSheet/disabled) it's supported everywhere except IE (support is unknown)

Right, but there are some extensions that completely disable Javascript execution. In that case dark mode would be the default. Consider adding `disabled="true"` here and changing it in JS.

stephanegigandet commented:
> > It seems like `prefers-color-scheme` requires user to activate dark theme on OS level. Correct me if I'm wrong. In this case I'm not sure if it's better

I don't know either. Is there some research on dark mode users? Do they turn it on or off depending on ambient light? Does it make sense to turn it on and off depending on the web sites users visit?

I was under the impression that it was more of a "global" preference that is valid accross sites and times: you either want dark mode all the time, or you don't. So you set it once in the OS preference, and then hopefully web sites have styles that match your preference.

GendelfLugansk commented:
> I was under the impression that it was more of a "global" preference that is valid accross sites and times: you either want dark mode all the time, or you don't. So you set it once in the OS preference, and then hopefully web sites have styles that match your preference.

The thing is, when I google "Chrome dark mode", I see [this](https://support.google.com/chrome/answer/9275525?co=GENIE.Platform%3DDesktop&hl=en&oco=0) article which implies that dark mode is supported only in Win 10+ and MacOS 10.14+. As Ubuntu 18.04 user, I don't think I can enable it. So for various Linux users toggle might be useful. Toggle can be combined with `prefers-color-scheme`, as was [suggested above](https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#issuecomment-623132902)

VaiTon commented:
> The thing is, when I google "Chrome dark mode", I see [this](https://support.google.com/chrome/answer/9275525?co=GENIE.Platform%3DDesktop&hl=en&oco=0) article which implies that dark mode is supported only in Win 10+ and MacOS 10.14+. As Ubuntu 18.04 user, I don't think I can enable it. So for various Linux users toggle might be useful. Toggle can be combined with `prefers-color-scheme`, as was [suggested above](https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#issuecomment-623132902)

I agree. Other sites (eg. stackoverflow, youtube, etc.) are using a site-specific switch and eventually we could add the "system settings" one.

alifafaruk commented:
Hi all! Thank you for taking the time to look into the pull request. My group (@Jen-Lopez and @lucasortizny) and I are new to web development and are unfamiliar with the mechanics of how to implement a prefers-color-scheme. We will be fixing the CSS to fix the bugs.

stephanegigandet commented:
Hi Everyone, just wondering if there is still some work being done / to do on this PR?

teolemon commented:
poke @alifafaruk @VaiTon

VaiTon commented:
> 
> 
> poke @alifafaruk @VaiTon

Cannot push to branch :)

alifafaruk commented:
Hi, we fixed all the style changes requested for dark mode. We also fixed bugs that were present. However, we don't know how to implement "prefers-color-scheme," so the method used to implement dark mode is still a toggle. If someone could help us change it to a "prefers-color-scheme" method, that would be great!

GendelfLugansk commented:
This is still an issue:
![screenshot](https://camo.githubusercontent.com/6a327708d668706d6a2380777bea3218ffc848c7/68747470733a2f2f73637265656e73686f74732e67656e6e6164792e70702e75612f32303230303530335f3536303830613862662e706e67)
"Android apk" is still unreadable:
![screenshot](https://camo.githubusercontent.com/6e89af72175c7be309b05bb119c0f8190f150469/68747470733a2f2f73637265656e73686f74732e67656e6e6164792e70702e75612f32303230303530335f3436386562386466372e706e67)
Not sure about this wide border around each tag (if others are ok with it, it's fine with me):
![screenshot](https://screenshots.gennady.pp.ua/20200601_8f035a41c.png)
This is still an issue (go to "Charts and maps", choose "Energy" for horizontal axis and click "generate graph"):
![screenshot](https://camo.githubusercontent.com/4a52cbf0865fb6539c617a32846de0d12d448eb9/68747470733a2f2f73637265656e73686f74732e67656e6e6164792e70702e75612f32303230303530335f6435633239346262312e706e67)

GendelfLugansk commented on lib/ProductOpener/Display.pm:
```diff
@@ -6090,7 +6090,13 @@ $options{favicons}
 <link rel="search" href="$formatted_subdomain/cgi/opensearch.pl" type="application/opensearchdescription+xml" title="$Lang{site_name}{$lang}">
 $header
+<!-- Dark Mode -->
+<link rel="stylesheet" href="$static_subdomain/css/darkMode.css">
```
Add `disabled="true"` to `link` tag

GendelfLugansk commented:
Please make these requested changes:

https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#pullrequestreview-404621626
https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#pullrequestreview-404634024
https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#pullrequestreview-422133026

Also, please look at these screenshots: https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#issuecomment-637064818

teolemon commented:
I created an APK logo in line with the others: 
https://github.com/openfoodfacts/openfoodfacts-server/pull/3535/files?short_path=67a224e#diff-67a224e98ae5f4a0802aeac7113905f3

Jen-Lopez commented:
Hi! Thank you all for the suggestions. I tried making the JS work but I also agree that some browsers may have javascript disabled and by default, dark mode would be enabled (which is something that shouldn't happen). I tried adding the disabled = "true" ([request](https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#pullrequestreview-422133026)) to the dark mode stylesheet but it didn't seem to work on my end (dark mode stopped working completely). I looked into the prefers-color-scheme and it seems that most apps and websites are using something similar to implement a dark mode feature nowadays. Although the prefers-color-scheme is not compatible with some browsers and is based on system preferences, I still think it's a good and efficient way for dark mode implementation without disrupting the original stylesheet (and with no glitches). @alifafaruk and I are still in the process of making [changes](https://github.com/openfoodfacts/openfoodfacts-server/pull/3317#issuecomment-637064818) but the recent commit implements the [prefers-color-scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) suggested previously, with JS removed.

teolemon commented:
Good news, the dark mode compatible APK logo is merged, and available in prod

Jen-Lopez commented:
I fixed the "Edit a Product" Page and went throughout the site to fix more css bugs. Since the new Android image is available in production and is dark mode aware, this issue should be resolved. However, I was unable to fix the floating WebLinks block in "dairies" since there is no id/class associated with it to change the styling. What page should I look at to add an ID to the element? Also, if there are any changes to be made or any suggestions, please let me know.

teolemon commented:
This PR has merge conflicts.

CloCkWeRX commented:
TODO items

- [ ] Android APK - Anyone able to do a revised version of https://static.openfoodfacts.org/images/misc/android-apk.svg ? 
- [ ] Weblinks - is this just https://github.com/openfoodfacts/openfoodfacts-server/blob/ec77a782631701af3db2aa799789214b7e5a2cce/lib/ProductOpener/Display.pm#L3255 with inline styles that should be pulled up to a main stylesheet?
- [ ] Tabs - ```ul.tabs {
    border-top: 1px solid #ccc;
    background-color: #efefef;
}```` ?

CloCkWeRX commented:
Did a dark friendly (and light friendly) revamp of the logo
![android-apk](https://user-images.githubusercontent.com/365751/159255209-3225ccfb-51a7-42c4-be03-1b8bd615ac28.svg)

teolemon commented:
We just deployed a new design, but without dark mode so far.

stephanegigandet commented:
We deployed a redesign of the https://world.openfoodfacts.org website, so sadly we won't be able to deploy this PR.
Sorry about that and thank you for your work @alifafaruk @lucasortizny and @Jen-Lopez .
